### PR TITLE
[ArgPromotion] Remove redundant logic from recursive argpromotion code

### PR DESCRIPTION
**Pull request by vedantparanjape-amd**
This patch further cleans up the implementation by removing some redundant checks and replacing cast<> with get() calls.

This contribution is based on the discussion in #78735

### CODE REVIEW

**Comment by vedantparanjape-amd** [comment]
@nikic @arsenm I have created this PR to address the changes discussed in the PR #78735

File: llvm/lib/Transforms/IPO/ArgumentPromotion.cpp [hunk 1]
```diff
@@ -640,16 +640,18 @@ static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
...
+            CB->getFunction()->getReturnType()) {
       if (PtrArg != Arg) {
         LLVM_DEBUG(dbgs() << "ArgPromotion of " << *Arg << " failed: "
                           << "pointer offset is not equal to zero\n");
         return false;
       }
 
       unsigned int ArgNo = Arg->getArgNo();
-      if (CB->getArgOperand(ArgNo) != Arg || U->getOperandNo() != ArgNo) {
+      if (U->getOperandNo() != ArgNo) {
```
Comment (arsenm) [inline llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:654]: Can you use isCallee?

File: llvm/lib/Transforms/IPO/ArgumentPromotion.cpp [hunk 2]
```diff
@@ -640,16 +640,18 @@ static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
     }
 
     auto *CB = dyn_cast<CallBase>(V);
-    Value *PtrArg = cast<Value>(U);
-    if (CB && PtrArg && CB->getCalledFunction() == CB->getFunction()) {
+    Value *PtrArg = U->get();
+    if (CB && CB->getCalledFunction() == CB->getFunction() &&
+        CB->getCalledFunction()->getReturnType() ==
+            CB->getFunction()->getReturnType()) {
```
Comment (arsenm) [inline llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:646]: Why is the function return type important? getCalledFunction is supposed to return null if the function types mismatch already?

Comment (vedantparanjape-amd) [reply llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:646]: I am confused, wasn't I meant to add the check to see if the function call type doesn't match the function def?

File: llvm/lib/Transforms/IPO/ArgumentPromotion.cpp (hunk 1 above)
Comment (vedantparanjape-amd) [reply llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:654]: I don't think so, as I need to check the position of the argument in the call matches the position in the caller function.

File: llvm/lib/Transforms/IPO/ArgumentPromotion.cpp (hunk 2 above)
Comment (nikic) [reply llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:646]: Ah sorry, I forgot that getCalledFunction already checks the function type. Ignore my comment about that.

Comment (vedantparanjape-amd) [reply llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:646]: okay, cool! so I can remove that check ? and then can I get a LGTM.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**